rascal-lsp/src/main/rascal/lang/rascal/lsp/refactor/rename/Common.rsc

```diff
@@ -53,22 +54,41 @@ data RenameConfig(
   , PathConfig(loc) getPathConfig = PathConfig(loc l) { throw "No path config for <l>"; }
 );
 
-bool(loc) containsFilter(type[&T <: Tree] t, str name, str(str) escape, Tree(loc) getTree) {
-    Tree n = parse(t, name);
-    Tree en = parse(t, escape(name));
-    return bool(loc l) {
-        bottom-up-break visit (getTree(l)) {
-            case n: return true;
-            case en: return true;
-        }
-        return false;
-    };
+bool(loc) containsFilter(type[&T <: Tree] T, str name, Tree(loc) getTree) {
+    if (just(t) := tryParseAs(T, name)) {
+        return bool(loc l) {
+            bottom-up-break visit (getTree(l)) {
+                case t: return true;
+            }
+            return false;
+        };
+    }
+
+    // If we can't parse the name at all, we don't even have to traverse the tree - we'll never find this.
+    return bool(loc _) { return false; };
 }
 
-tuple[set[loc], set[loc]] findOccurrenceFilesSymmetric(type[&T <: Tree] N, str name, set[loc]() getSourceFiles, Tree(loc) getTree) {
-    containsName = containsFilter(N, name, rascalEscapeName, getTree);
-    set[loc] fs = {f | loc f <- getSourceFiles(), containsName(f)};
-    return <fs, fs>;
+bool isContainedInScope(loc l, loc scope, TModel tm) {
+    // lexical containment
+    if (isContainedIn(l, scope)) return true;
+
+    // via import/extend
+    set[loc] reachableFrom = (tm.paths<pathRole, to, from>)[{importPath(), extendPath()}, scope];
+    return any(loc fromScope <- reachableFrom, isContainedIn(l, fromScope));
+}
+
+set[loc] findNameOccurrenceFiles(str name, set[loc] sourceFiles, Tree(loc) getTree) {
+    containsName = containsFilter(#Name, name, getTree);
+    containsQName = containsFilter(#QualifiedName, name, getTree);
+    containsNonTerm = containsFilter(#Nonterminal, name, getTree);
+    containsNonTermLabel = containsFilter(#NonterminalLabel, name, getTree);
+
+    return {f | loc f <- sourceFiles, containsName(f) || containsQName(f) || containsNonTerm(f) || containsNonTermLabel(f)};
```

Comment:
it's a pitty that these `containsFilter` functions how all have to iterate over the same tree.

In a way it's more composable this way, but also, it could have been a single visit.

Reply:
Indeed. This has now been fixed.